contrib/babelfishpg_tsql/src/pltsql_ruleutils.c

```diff
@@ -0,0 +1,2414 @@
+	if (!HeapTupleIsValid(tuple))
+	{
+		if ((flags & FORMAT_TYPE_ALLOW_INVALID) != 0)
+			return pstrdup("???");
+		else
+			elog(ERROR, "cache lookup failed for type %u", type_oid);
+	}
+	typeform = (Form_pg_type) GETSTRUCT(tuple);
+
+	with_typemod = (flags & FORMAT_TYPE_TYPEMOD_GIVEN) != 0 && (typemod >= 0);
+
+	buf = NULL;
+	LOCAL_FCINFO(fcinfo, 1);
+
+	InitFunctionCallInfoData(*fcinfo, NULL, 0, InvalidOid, NULL, NULL);
+	fcinfo->args[0].value = ObjectIdGetDatum(type_oid);
+	fcinfo->args[0].isnull = false;
+	tsql_typename = (*translate_pg_type_to_tsql) (fcinfo);
+
+	/*
+	 * If it is TSQL type then report it without any qualification.
+	 */
+	if (tsql_typename)
+		buf = text_to_cstring(DatumGetTextPP(tsql_typename));
+
+	if (buf == NULL)
+	{
+		/*
+		 * Default handling: report the name as it appears in the catalog.
+		 * Here, we must qualify the name if it is not visible in the search
+		 * path or if caller requests it; and we must double-quote it if it's
+		 * not a standard identifier or if it matches any keyword.
+		 */
+		char	   *nspname;
+		char	   *typname;
+
+		if ((flags & FORMAT_TYPE_FORCE_QUALIFY) == 0 &&
+			TypeIsVisible(type_oid))
+			nspname = NULL;
+		else
+			nspname = get_namespace_name_or_temp(typeform->typnamespace);
+
+		typname = NameStr(typeform->typname);
+
+		buf = quote_qualified_identifier(nspname, typname);
+	}
+
+	if (with_typemod)
+	{
+		buf = tsql_printTypmod(buf, typemod, typeform->typmodout);
+	}
+
+	ReleaseSysCache(tuple);
+
+	return buf;
+}
+
+/*
+ * Add typmod decoration to the basic type name
+ */
+static char *
+tsql_printTypmod(const char *typname, int32 typmod, Oid typmodout)
+{
+	char	   *res;
+
+	/* Shouldn't be called if typmod is -1 */
+	Assert(typmod >= 0);
+
+	/*
+	* In case of time, datetime2 or datetimeoffset print typmod
+	* info directly because it uses timestamp typmodout function
+	* which appends timezone data along with typmod which is not
+	* required. Directly print typename for smalldatetime as it
+	* doesn't support typmod.
+	*/
+	if (strcmp(typname, "time") == 0 ||
+	   strcmp(typname, "datetime2") == 0 ||
+	   strcmp(typname, "datetimeoffset") == 0)
+	{
+		res = psprintf("%s(%d)", typname, (int) typmod);
+	}
+	else if (strcmp(typname, "smalldatetime") == 0)
+	{
+		res = pstrdup(typname);
+	}
+	else
+	{
+		if (typmodout == InvalidOid)
+		{
+			/* Default behavior: just print the integer typmod with parens */
+			res = psprintf("%s(%d)", typname, (int) typmod);
+		}
+		else
+		{
+			/* Use the type-specific typmodout procedure */
+			char	   *tmstr;
+
+			tmstr = DatumGetCString(OidFunctionCall1(typmodout,
+													 Int32GetDatum(typmod)));
+			res = psprintf("%s%s", typname, tmstr);
+		}
+	}
+	return res;
+}
+
+/*
+ * This version allows a nondefault typemod to be specified.
+ */
+static char *
+tsql_format_type_with_typemod(Oid type_oid, int32 typemod)
+{
+	return tsql_format_type_extended(type_oid, typemod, FORMAT_TYPE_TYPEMOD_GIVEN);
```

Comment:
We can directly call tsql_format_type_extended( <>, <>, FORMAT_TYPE_TYPEMOD_GIVEN);
What is the purpose of this level of encapsulation when the function is defined in the same file?